### PR TITLE
Instrumenting follow-up: metrics labelling w/ finer granularity

### PR DESCRIPTION
**Pull request by yue9944882**
the new classes are picked from https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/651, except for those groovy unit tests..

a sample output from the prometheus example will be:

```
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.005",} 0.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.01",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.025",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.05",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.075",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.1",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.25",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.5",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="0.75",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="1.0",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="2.5",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="5.0",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="7.5",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="10.0",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",le="+Inf",} 2.0
k8s_java_resource_request_latency_seconds_count{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",} 2.0
k8s_java_resource_request_latency_seconds_sum{http_response_code="404",api_group="",api_version="v1",resource="pods",subresource="",api_verb="get",namespace="bar",} 0.014409807
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.005",} 0.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.01",} 0.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.025",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.05",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.075",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.1",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.25",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.5",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="0.75",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="1.0",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="2.5",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="5.0",} 1.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="7.5",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="10.0",} 2.0
k8s_java_resource_request_latency_seconds_bucket{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",le="+Inf",} 2.0
k8s_java_resource_request_latency_seconds_count{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",} 2.0
k8s_java_resource_request_latency_seconds_sum{http_response_code="200",api_group="",api_version="v1",resource="pods",subresource="",api_verb="list",namespace="",} 5.198464195000001
```

### CODE REVIEW

**Comment by brendandburns** [comment]
I don't think we want to remove the overall histogram for request latencies, given the fine grained histograms it's impossible to re-assemble the broader histograms.

Likewise, it's a little weird to write a query like rate(k8s_java_resource_request_latency_seconds_count{code="500"}) to get the rate of errors. I get that we might not want to have extra Counter metrics, but I feel like it makes more sense than overloading the histogram metrics.

So I would retain both of the Counters and also the all-up Histogram.

The rest of the additional labels LGTM.

**Comment by yue9944882** [comment]
> given the fine grained histograms it's impossible to re-assemble the broader histograms.

@brendandburns if i understand correctly, we can always aggregate finer-grained histogram/counter to overall histogram/counter by promQL functions --- generally the fine-grained metrics is a superset of overall metrics:

- >   static final Counter requests =	
      Counter.build().name(PREFIX + "_requests_total").help("Total requests").register();

  will be equal to
```
sum(rate(k8s_java_resource_request_latency_seconds_sum[10m]) by (code)
```

- > Histogram.build()	      Histogram.build()
          .name(PREFIX + "_request_latency_seconds")

  can be aggregated by

```
histogram_quantile(0.9, sum (rate(k8s_java_resource_request_latency_seconds_bucket[10m]) by (le)))
```

am i missing anything?

**Comment by brendandburns** [comment]
a) I believe that the first calculation is a histogram of the average latency, not the actual latency, so it's lossy.

b) The second calculation is valid, but I think it's hard for a person to understand, since why would I sum latency count to get # of requests. So while the calculation works, I think it adds complexity.

**Comment by yue9944882** [comment]
> a) I believe that the first calculation is a histogram of the average latency, not the actual latency, so it's lossy.

oh oversight! i was referring to `k8s_java_resource_request_latency_seconds_count` to replace the counter.

> b) The second calculation is valid, but I think it's hard for a person to understand, since why would I sum latency count to get # of requests. So while the calculation works, I think it adds complexity.

yeah i agree the aggregation is complex and not intuitive enough, even tho they're overlapping mathematically. will keep counter and all-up histogram

**Comment by yue9944882** [comment]
@brendandburns i added the request-counter and response-code-counter back, PTAL

File: util/src/main/java/io/kubernetes/client/monitoring/PrometheusInterceptor.java [hunk 1]
```diff
@@ -12,39 +12,85 @@
...
-      codes.labels(Integer.toString(response.code())).inc();
-      return response;
+      response = chain.proceed(chain.request());
     } finally {
-      timer.observeDuration();
+      timecost = timer.elapsedSeconds();
+    }
+
+    codes.labels(Integer.toString(response.code())).inc();
+    ;
```
[the remark this answers is not in the document]
Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/monitoring/PrometheusInterceptor.java]: done

File: util/src/main/java/io/kubernetes/client/apimachinery/KubernetesResource.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      String name) {
+    this.apiGroup = apiGroup;
+    this.apiVersion = apiVersion;
+    this.resource = resource;
+    this.subResource = subResource;
+    this.namespace = namespace;
+    this.name = name;
+  }
+
+  private final String apiGroup;
```
[the remark this answers is not in the document]
Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/apimachinery/KubernetesResource.java]: i did the TODO right now

**Pull request merged**